File: Dopamine.Controls/DataGridEx.vb
''' <summary>
''' DataGrid which propagates its DataContext to its Columns
''' </summary>
''' <remarks></remarks>
Public Class DataGridEx
    Inherits DataGrid

    Protected Overrides Sub OnInitialized(e As EventArgs)
        MyBase.OnInitialized(e)

        If Me.DataContext IsNot Nothing AndAlso Me.Columns.Count > 0 Then
            For Each col As DataGridColumn In Me.Columns
                col.SetValue(FrameworkElement.DataContextProperty, Me.DataContext)
            Next
        End If
    End Sub
End Class

File: Dopamine.Presentation/Controls/DataGridEx.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace Dopamine.Presentation.Controls
{
    public class DataGridEx : DataGrid
    {
        protected override void OnInitialized(EventArgs e)
        {
            base.OnInitialized(e);

            if (this.DataContext != null && this.Columns.Count > 0)
            {
                foreach (DataGridColumn col in this.Columns)
                {
                    col.SetValue(FrameworkElement.DataContextProperty, this.DataContext);
                }
            }
        }
    }
}
